File: src/Scripting/Ruby/patterns/presentation_abstraction_control.rb
# frozen_string_literal: true

module RubyPatterns
  module PresentationAbstractionControl
    module_function

    def run
      state = { value: 1 }
      control = ->(delta) { state[:value] += delta }
      presentation = -> { state[:value].to_s }
      control.call(2)
      raise 'Presentation-Abstraction-Control failed' unless presentation.call == '3'
      true
    end
  end
end

RubyPatterns::PresentationAbstractionControl.run if $PROGRAM_NAME == __FILE__

File: src/Scripting/Lua/patterns/presentation_abstraction_control.lua
local abstraction = { value = 1 }
local control = { increment = function(a) a.value = a.value + 1 end }
local presentation = { render = function(a) return "value=" .. a.value end }
control.increment(abstraction)
assert(presentation.render(abstraction) == "value=2")
return true
